src/jupyterlabcontroller/services/prepuller.py

```diff
@@ -0,0 +1,168 @@
+            return
+        self._logger.info("Starting prepuller tasks")
+
+        # Wait for the image data to populate in the foreground. This ensures
+        # that we populate image data before FastAPI completes its startup
+        # event, and therefore before we start answering requests. That in
+        # turn means a more accurate health check, since until we have
+        # populated image data our API is fairly useless. (It also makes life
+        # easier for the test suite.)
+        await self._image_service.prepuller_wait()
+        await self._scheduler.spawn(self._prepull_loop())
+        self._running = True
+
+    async def stop(self) -> None:
+        if not self._running:
+            self._logger.info("Prepuller tasks were already stopped")
+        self._logger.info("Cancelling prepuller tasks")
+        await self._scheduler.close()
+        self._running = False
+
+    async def _prepull_loop(self) -> None:
+        """Continually prepull images in a loop.
+
+        When the prepuller is stopped, we will orphan prepuller pods. Avoiding
+        this is difficult and unreliable. The prepuller should instead detect
+        orphaned pods on startup and clean them up.
+        """
+        while True:
+            await self.prepull_images()
+            await self._image_service.prepuller_wait()
+
+    async def prepull_images(self) -> None:
+        """Prepull missing images."""
+        missing_by_node = self._image_service.missing_images_by_node()
+
+        # Try to avoid hammering nodes by prepulling at most one image per
+        # node at a time. We do this by creating a separate background task
+        # per node, each of which works through the list of images that are
+        # missing on that node.
+        node_tasks = set()
+        for node, images in missing_by_node.items():
+            self._logger.debug(f"Creating prepull task for {node}")
+            prepull_call = self._prepull_images_for_node(node, images)
+            task = asyncio.create_task(prepull_call)
+            node_tasks.add(task)
+        await asyncio.gather(*node_tasks)
+        self._logger.debug("Finished prepulling all images")
+
+    def _prepull_pod_spec(self, image: RSPImage, node: str) -> V1PodSpec:
+        """Create a spec to run a pod with a specific image on a node.
+
+        The pod does nothing but sleep five seconds and then exit.  Its only
+        function is to ensure that that image gets pulled to that node.
+        """
+        return V1PodSpec(
+            containers=[
+                V1Container(
+                    name="prepull",
+                    command=["/bin/sleep", "5"],
+                    image=image.reference_with_digest,
+                    working_dir="/tmp",
+                )
+            ],
+            node_name=node,
+            restart_policy="Never",
+        )
+
+    async def _prepull_image(self, image: RSPImage, node: str) -> None:
```

Comment:
Maybe put the last 8 chars of the nodename in the pod name too?  It's nice to be able to see without having to dig through the YAML which node a misbehaving prepull pod should be on.

Reply:
Ack, sorry, we talked about this and I forgot that I committed something I was playing around with and didn't go back and fix it. Should now be fixed properly based on the Kubernetes object naming rules.